plugins/imfile/imfile.c

```diff
@@ -2522,6 +2562,204 @@ do_inotify(void)
+
+	DBGPRINTF("imfile: fen_processEvent: %s\n", fobjp->fo_name);
+
+	/* Port needs to be reassociated */
+	pLstn->bPortAssociated = 0;
+
+	/* Get File Stats */
+	if (!(revents & FILE_EXCEPTION) && stat(fobjp->fo_name, &statFile) == -1) {
+		DBGPRINTF("imfile: fen_processEvent: Failed to stat file "
+			": %s - errno %d\n", fobjp->fo_name, errno);
+		ABORT_FINALIZE(RS_RET_FILE_NO_STAT);
+	}
+
+	/*
+	* Add what ever processing that needs to be done
+	* here. Process received events.
+	*/
+	if (revents) {
+		fen_printevent(revents, fobjp->fo_name);
+
+		/* File has been modified, trigger a pollFile */
+		if (revents & FILE_MODIFIED) {
+			pollFile(pLstn, NULL);
+		}
+
+		/* If exception, no need to re-register. */
+		if (revents & FILE_EXCEPTION) {
+			free(pLstn->pfinf->fobj.fo_name);
+			free(pLstn->pfinf);
+			pLstn->pfinf = NULL;
+			ABORT_FINALIZE(RS_RET_SYS_ERR);
+		}
+	}
+
+	/* Register file event */
+	fobjp->fo_atime = statFile.st_atim;
+	fobjp->fo_mtime = statFile.st_mtim;
+	fobjp->fo_ctime = statFile.st_ctim;
+	DBGPRINTF("imfile: fen_processEvent: associate port for %s\n", fobjp->fo_name);
+	if (port_associate(port, PORT_SOURCE_FILE, (uintptr_t)fobjp,
+				pLstn->pfinf->events, (void *)pLstn) == -1) {
+		/* Add error processing as required, file may have been deleted/moved. */
+		errmsg.LogError(1, RS_RET_SYS_ERR, "imfile: fen_processEvent: Failed to register file "
+			": %s - errno %d\n", fobjp->fo_name, errno);
+		free(pLstn->pfinf->fobj.fo_name);
+		free(pLstn->pfinf);
+		pLstn->pfinf = NULL;
+		ABORT_FINALIZE(RS_RET_SYS_ERR);
+	} else {
+		/* Port successfull listening now*/
+		pLstn->bPortAssociated = 1;
+	}
+
+finalize_it:
+	RETiRet;
+}
+
+static rsRetVal
+do_fen(void)
+{
+	int port;
+	int bHadFileData; /* were there at least one file with data during this run? */
+	port_event_t portEvent;
+	struct timespec timeout;
+	lstn_t *pLstn;	/* Listener helper*/
+	DEFiRet;
+
+	/* Set port timeout to 1 second. We need to checkfor unmonitored files during meantime */
+	timeout.tv_sec = 1;
+	timeout.tv_nsec = 0;
+
+	/* create port instance */
+	if ((port = port_create()) == -1) {
+		errmsg.LogError(1, RS_RET_INOTIFY_INIT_FAILED, "imfile: do_fen: Init port failed ");
+		return RS_RET_INOTIFY_INIT_FAILED;
+	}
+
+	/* Loop through all configured listeners */
+	for(pLstn = runModConf->pRootLstn ; pLstn != NULL ; pLstn = pLstn->next) {
+		if(pLstn->masterLstn == NULL) {
+			// Create FileInfo struct
+			pLstn->pfinf = malloc(sizeof(struct fileinfo));
+			if (pLstn->pfinf == NULL) {
+				errmsg.LogError(1, RS_RET_INOTIFY_INIT_FAILED, "imfile: do_fen: alloc memory "
+					"for fileinfo failed ");
+				ABORT_FINALIZE(RS_RET_INOTIFY_INIT_FAILED);
+			}
+
+			if ((pLstn->pfinf->fobj.fo_name = strdup((char*)pLstn->pszFileName)) == NULL) {
+				errmsg.LogError(1, RS_RET_INOTIFY_INIT_FAILED, "imfile: do_fen: alloc memory "
+					"for strdup failed ");
+				free(pLstn->pfinf);
+				pLstn->pfinf = NULL;
+				ABORT_FINALIZE(RS_RET_INOTIFY_INIT_FAILED);
```

Comment:
shouldn't we better use a FEN-specific error code? I mean this might be easier in troubleshooting...

Reply:
Yes thats better, I added specific error code RS_RET_FEN_INIT_FAILED. Tests are running